eng/pipelines/templates/stages/vmr-source-build.yml

```diff
@@ -0,0 +1,204 @@
+### If run in a PR, new changes are applied to a local copy of the VMR, then it is source-built and tested
+
+parameters:
+  # Resource id of the installer build to retrieve source/product tarball's from
+  installerBuildResourceId: current
+
+  # Branch of the VMR to use (to push to for internal builds)
+  vmrBranch: $(Build.SourceBranchName)
+
+  # The following parameters aren't expected to be passed in rather they are used for encapsulation
+  # -----------------------------------------------------------------------------------------------
+  centOSStream8Container: mcr.microsoft.com/dotnet-buildtools/prereqs:centos-stream8
+  centOSStream9Container: mcr.microsoft.com/dotnet-buildtools/prereqs:centos-stream9
+  debian11Arm64Container: mcr.microsoft.com/dotnet-buildtools/prereqs:debian-11-arm64v8
+  fedora36Container: mcr.microsoft.com/dotnet-buildtools/prereqs:fedora-36
+  ubuntu2004Container: mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-20.04
+
+  # Internal builds
+  poolInternalAmd64:
+    name: NetCore1ESPool-Svc-Internal
+    demands: ImageOverride -equals Build.Ubuntu.1804.Amd64
+  poolInternalAmd64PR:
+    name: NetCore1ESPool-Internal-XL
+    demands: ImageOverride -equals Build.Ubuntu.1804.Amd64
+  poolInternalArm64:
+    name: Docker-Linux-Arm-Internal
+
+  # Public builds / PRs
+  poolPublicAmd64:
+    name: NetCore-Public-XL
+    demands: ImageOverride -equals Build.Ubuntu.1804.Amd64.Open
+
+stages:
+- stage: VMR_Source_Build
+  displayName: VMR Source-Build
+  ${{ if eq(parameters.installerBuildResourceId, 'current') }}:
+    dependsOn:
+    - Build_Linux
+    - Synchronize_VMR
+
+  variables:
+  - ${{ if eq(parameters.installerBuildResourceId, 'current') }}:
+    - name: vmrIsPushed
+      value: $[stageDependencies.Synchronize_VMR.Synchronize_VMR.outputs['vmrPush.pushed']]
+    - name: vmrRevision
+      value: $[stageDependencies.Synchronize_VMR.Synchronize_VMR.outputs['vmrPush.sha']]
+  - ${{ else }}:
+    - name: vmrIsPushed
+      value: true
+    - name: vmrRevision
+      value: ''
+  - ${{ if eq(variables['System.TeamProject'], 'public') }}:
+    - name: defaultPoolName
+      value: ${{ parameters.poolPublicAmd64.name }}
+    - name: defaultPoolDemands
+      value: ${{ parameters.poolPublicAmd64.demands }}
+  - ${{ if eq(variables['System.TeamProject'], 'internal') }}:
+    - ${{ if in(variables['Build.Reason'], 'PullRequest') }}:
+      - name: defaultPoolName
+        value: ${{ parameters.poolInternalAmd64PR.name }}
+      - name: defaultPoolDemands
+        value: ${{ parameters.poolInternalAmd64PR.demands }}
+    - ${{ else }}:
+      - name: defaultPoolName
+        value: ${{ parameters.poolInternalAmd64.name }}
+      - name: defaultPoolDemands
+        value: ${{ parameters.poolInternalAmd64.demands }}
+
+  jobs:
+  - template: ../jobs/vmr-source-build.yml
+    parameters:
+      name: CentOSStream8_Online
+      installerBuildResourceId: ${{ parameters.installerBuildResourceId }}
+      vmrUpdate:
```

Comment:
Yeah, this is a step back from the matrix solution where we could unify this into one place. I also had a version where I pulled these one level up and sent them down as object parameters but it didn't quite work because I need to call this from two different places.